lerobot/common/utils/benchmark.py

```diff
@@ -0,0 +1,106 @@
+#!/usr/bin/env python
+
+# Copyright 2024 The HuggingFace Inc. team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import threading
+import time
+from contextlib import ContextDecorator
+from functools import wraps
+
+
+class TimeBenchmark(ContextDecorator):
+    """
+    Measures execution time using a context manager or decorator.
+
+    This class supports both context manager and decorator usage, and is thread-safe for multithreaded
+    environments.
+
+    Args:
+        print: If True, prints the elapsed time upon exiting the context or completing the function. Defaults
+        to False.
+
+    Usage examples:
+
+        benchmark = TimeBenchmark()
+
+        # As a context manager
+        with benchmark:
+            time.sleep(1)
+        print(f"Block took {benchmark.result:.4f} seconds")
+
+        # As a decorator
+        @benchmark
+        def example_function():
+            time.sleep(1)
+            return "Function result"
+
+        result, elapsed_time = example_function()
+        print(f"Function took {elapsed_time:.4f} seconds and returned '{result}'")
+
+        # With multithreading
+        import threading
+
+        def context_manager_example():
+            with benchmark:
+                time.sleep(1)
+            print(f"Block took {benchmark.result:.4f} seconds")
+
+        def decorator_example():
+            result, elapsed_time = example_function()
+            print(f"Function took {elapsed_time:.4f} seconds and returned '{result}'")
+
+        threads = []
+        for _ in range(3):
+            t1 = threading.Thread(target=context_manager_example)
+            t2 = threading.Thread(target=decorator_example)
+            threads.extend([t1, t2])
+
+        for t in threads:
+            t.start()
+
+        for t in threads:
+            t.join()
```

Comment:
Add "```python"

Reply:
Could you create multiple python blocks and display what is the expected results ?

Reply:
Overall I feel this class is overkill. Why not use:
```python
start_time_s = time.perf_counter()
elapsed_time_s = time.perf_counter() - start_time_s 
```
Maybe I am missing something

Reply:
This is just a convenience to make code more readable (and additionally handle threaded timing) and turn this code:
```python
start_time_s = time.monotonic()
frames = decode_frames_fn(
    video_path, timestamps=timestamps, tolerance_s=1e-4, backend=backend, **decoder_kwgs
)
avg_load_time = (time.monotonic() - start_time_s) / num_frames
list_avg_load_time.append(avg_load_time)

start_time_s = time.monotonic()
original_frames = load_original_frames(imgs_dir, timestamps)
avg_load_time_from_images = (time.monotonic() - start_time_s) / num_frames
list_avg_load_time_from_images.append(avg_load_time_from_images)
```
into this:
```python
with time_benchmark:
    frames = decode_video_frames(video_path, timestamps=timestamps, tolerance_s=5e-1, backend=backend)
result["load_time_video_ms"] = time_benchmark.result_ms / num_frames

with time_benchmark:
    original_frames = load_original_frames(imgs_dir, timestamps, fps)
result["load_time_images_ms"] = time_benchmark.result_ms / num_frames
```

I initially wanted it to be able to be used both as a context manager and a decorator, but I only need the context manager and I doubt the decorator part of that class will ever be useful so I can remove that, which will simplify things.

Reply:
> Could you create multiple python blocks and display what is the expected results ?

Fixed https://github.com/huggingface/lerobot/pull/282/commits/74019483f324ad9cc57b4fbb7f3b5880471bdf40